Restore flags to include from file and exclude from file

konidev20 opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
Added `--exclude-file`, `--iexclude-file`, `include-file` and `iinclude-file` to the restore command. 

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Closes #4781 

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].
-->

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

konidev20 committed:
restore: read includes, insensitive includes, excludes and insensitive excludes from a file

feature for gh-4781

konidev20 commented:
@MichaelEischer please review this and let me know if I am on the right track.

konidev20 committed:
fix: add string.Lower for insenstive includes and excludes read from file

MichaelEischer commented on cmd/restic/cmd_restore.go:
```diff
@@ -176,6 +184,31 @@ func runRestore(ctx context.Context, opts RestoreOptions, gopts GlobalOptions,
 	excludePatterns := filter.ParsePatterns(opts.Exclude)
 	insensitiveExcludePatterns := filter.ParsePatterns(opts.InsensitiveExclude)
+
+	if len(opts.ExcludeFiles) > 0 {
```
At least the exclude patterns look like we could share the implementation used by the `backup` command. Please remove the custom exclude pattern implementation from restore and use `excludePatternOptions` instead. Unless I've missed something, the returned `RejectByNameFunc`s should suitable for use in selectExcludeFilter.

konidev20 replied:
This is done.

MichaelEischer commented on cmd/restic/cmd_restore.go:
```diff
@@ -199,6 +232,31 @@ func runRestore(ctx context.Context, opts RestoreOptions, gopts GlobalOptions,
 	includePatterns := filter.ParsePatterns(opts.Include)
 	insensitiveIncludePatterns := filter.ParsePatterns(opts.InsensitiveInclude)
+
+	if len(opts.IncludeFiles) > 0 {
```
Please create a `includePatternOptions` struct similar to `excludePatternOptions`. The `includeByNameFunc`s must be somewhat different in that it has to use `filter.ListWithChild(includePatterns, item)` and return `matched, childMayMatch, err`.

konidev20 replied:
This is done. I added a file called `include.go`

MichaelEischer reviewed (commented):
I'd prefer a different design. `runRestore` is already too long, so I'd like to extract the include/exclude pattern parsing code. Judging from a quick look at the exclude code used by the backup command (see `cmd/restic/exclude.go`), it should be possible to directly reuse the code for the exclude options. And it should be possible to adapt that pattern to add new include options. (Feel free to ask if you need more details to understand what I have in mind)

konidev20 commented:
> I'd prefer a different design. `runRestore` is already too long, so I'd like to extract the include/exclude pattern parsing code. Judging from a quick look at the exclude code used by the backup command (see `cmd/restic/exclude.go`), it should be possible to directly reuse the code for the exclude options. And it should be possible to adapt that pattern to add new include options. (Feel free to ask if you need more details to understand what I have in mind)

I wanted to propose this too. I will update this draft over the next weekend.

konidev20 committed:
restore: refactor include and exclude

- added includePatternOptions similar to excludePatternOptions
- followed similar approach to backup for selecting files for restore

konidev20 committed:
fix: move include and exclude pattern validations to top

konidev20 committed:
test: add tests for reading patterns from file

MichaelEischer commented on cmd/restic/cmd_restore.go:
```diff
@@ -174,44 +149,28 @@ func runRestore(ctx context.Context, opts RestoreOptions, gopts GlobalOptions,
-			msg.E("error for iexclude pattern: %v", err)
-		}
+		for _, rejectFn := range excludePatternFns {
+			matched := rejectFn(item)
```
That logic is partially broken. If any `rejectFn` returns true then the file must be excluded. Right now, the loop only ever looks at the first function in `excludePatternFns`

You have to combine the results as follows:
```go
matched := false
for _, rejectFn := range excludePatternFns {
    matched = matched || rejectFn(item)
}
```

konidev20 replied:
My thought process was that I must exit the first match I get. Because it's an || operation. But I introduced an issue. I have resolved this.

konidev20 replied:
Let me know if that makes sense if we break after the first pattern it matches, it will be much faster. We don't have to do redundant pattern matching with other patterns.

MichaelEischer replied:
A short-circuit semantic sounds reasonable. That also works for the include filter, although there the optimization is only possible if both `selectedForRestore` and `childMayBeSelected` are true within the loop.

konidev20 replied:
@MichaelEischer I have made the changes, please check if it is correct.

MichaelEischer commented on cmd/restic/cmd_restore.go:
```diff
@@ -174,44 +149,28 @@ func runRestore(ctx context.Context, opts RestoreOptions, gopts GlobalOptions,
-		if err != nil {
-			msg.E("error for iexclude pattern: %v", err)
+		for _, includeFn := range includePatternFns {
+			selectedForRestore, childMayBeSelected = includeFn(item)
```
Same here, except that only the last function has an effect on the outcome.

konidev20 replied:
This is resolved. Please review.

konidev20 committed:
test: add tests for include By pattern

konidev20 committed:
changelog: update changelog for gh-4781

konidev20 committed:
doc: update documentation for restore command

konidev20 committed:
restore: accumulate results of multiple pattern checks

addressing review comments

konidev20 committed:
test: update test case name

MichaelEischer commented on cmd/restic/cmd_restore_integration_test.go:
```diff
@@ -51,13 +51,64 @@ func testRunRestoreLatest(t testing.TB, gopts GlobalOptions, dir string, paths [
 	rtest.OK(t, testRunRestoreAssumeFailure(snapshotID.String(), opts, gopts))
 }
 
+func TestRestoreIncludes(t *testing.T) {
```
Please extend the tests a bit to run restore at least once with a include or exclude file

konidev20 replied:
Added Unit tests for both.

MichaelEischer reviewed (changes requested):
I've found a few more issues. Currently `restore --exclude-file ...` does not work as `hasExcludes` is no longer correct. The same applies for include files.

konidev20 committed:
fix: include and exclude logic, add tests for include file and exclude file

konidev20 commented on cmd/restic/include.go:
```diff
@@ -0,0 +1,100 @@
+	f.StringArrayVarP(&opts.Includes, "include", "i", nil, "include a `pattern` (can be specified multiple times)")
+	f.StringArrayVar(&opts.InsensitiveIncludes, "iinclude", nil, "same as --include `pattern` but ignores the casing of filenames")
+	f.StringArrayVar(&opts.IncludeFiles, "include-file", nil, "read include patterns from a `file` (can be specified multiple times)")
+	f.StringArrayVar(&opts.InsensitiveIncludeFiles, "iinclude-file", nil, "same as --include-file but ignores casing of `file`names in patterns")
```
Done.

MichaelEischer replied:
Hmm, I just noticed that 
```
`file`names`
```
is actually correct. Without the backticks the help output changes from `--iinclude-file file` to `--iinclude-file string`. Please revert that change, sorry.

konidev20 replied:
Reverted.

konidev20 commented on cmd/restic/cmd_restore.go:
```diff
@@ -75,37 +70,17 @@ func init() {
-			return errors.Fatalf("--iinclude: %s", err)
-		}
-	}
+	hasExcludes := len(opts.Excludes) > 0 || len(opts.InsensitiveExcludes) > 0
```
I have added it as 

```
	hasExcludes := len(excludePatternFns) > 0
	hasIncludes := len(includePatternFns) > 0
```

Since, the length checks are done within the `CollectPatterns` functions. I have added a unit test as well to handle this exclusive check.

konidev20 commented on cmd/restic/integration_filter_pattern_test.go:
```diff
@@ -70,30 +70,85 @@ func TestRestoreFailsWhenUsingInvalidPatterns(t *testing.T) {
+
+	// Create an include file with some invalid patterns
+	includeFile := env.base + "/includefile"
+	fileErr := os.WriteFile(includeFile, []byte("*.go\n*[._]log[.-][0-9]\n!*[._]log[.-][0-9]"), 0644)
```
Done.

konidev20 commented:
Tests failing is unrelated.

MichaelEischer commented on cmd/restic/cmd_restore.go:
```diff
@@ -171,9 +178,12 @@ func runRestore(ctx context.Context, opts RestoreOptions, gopts GlobalOptions,
+			childMayBeSelected = childMayBeSelected && node.Type == "dir"
 
-		childMayBeSelected = childMayBeSelected && node.Type == "dir"
+			if selectedForRestore || childMayBeSelected {
```
I think that is wrong. If the first `includeFn` returns `false, true` (matched, childMayMatch) for a directory, then the `if` would exit the loop right now. However, a second `includeFn` could still return `true, true`. 

To handle that case correctly, the loop should look like the following:
```Go
for _, includeFn := range includePatternFns {
			matched, childMayMatch := includeFn(item)
			selectedForRestore = selectedForRestore || matched
			childMayBeSelected = childMayBeSelected || childMayMatch

			if selectedForRestore && childMayBeSelected {
				break
			}
		}
		childMayBeSelected = childMayBeSelected && node.Type == "dir"
```

konidev20 replied:
Done.

MichaelEischer commented on cmd/restic/cmd_restore_integration_test.go:
```diff
@@ -58,6 +59,67 @@ func testRunRestoreIncludes(t testing.TB, gopts GlobalOptions, dir string, snaps
+
+	opts := BackupOptions{}
+	// Perform backup
+	testRunBackup(t, filepath.Dir(env.testdata), []string{filepath.Base(env.testdata)}, opts, env.gopts)
```
The include/exclude check happens before even opening the repository, thus please remove the code that create a snapshot and just pass a placeholder value to `testRunRestoreAssumeFailure`. To make sure that we still get the correct error, let the `rtest.Assert` check that the error message contains the substring `exclude and include patterns are mutually exclusive`.

MichaelEischer commented on cmd/restic/cmd_restore_integration_test.go:
```diff
@@ -144,19 +223,38 @@ func TestRestoreFilter(t *testing.T) {
+
+	testRunRestoreExcludesFromFile(t, env.gopts, restoredir, snapshotID, patternsFile)
+
+	testRestoreFileExclusions(t, env, excludePatterns)
```
This call performs multiple restores of its own, but does not use the exclude file.

konidev20 replied:
Fixed, I made the logic simpler and similar to the inclusion tests.

MichaelEischer reviewed (commented):
I have a few more comment, after that we should be ready to go.

konidev20 committed:
fix: restore inclusion logic and restore tests
doc: update exclude and include docs

MichaelEischer reviewed (approved):
LGTM. Thanks for the blazingly fast fixes!
